activerecord/lib/active_record/associations/collection_association.rb

```diff
@@ -516,6 +521,11 @@ def find_by_scan(*args)
             load_target.select { |r| ids.include?(r.id.to_s) }
           end
         end
+
+        def index_in_target(record)
+          @target_index_map ||= @target.each_with_object({}).with_index { |(r, memo), i| memo[r] ||= i }
```

Comment:
```suggestion
          @target_index_map ||= @target.each_with_index.to_h
```

Reply:
I left a comment about this:
> The code this was replacing was `@target.index(record)` which would return the first index for duplicate records so this implementation keeps that behavior.

Do you think that change is meaningful? I definitely agree your suggested change is nicer and had actually initially implemented it like this before noticing the subtle behavior change. I'll plan commit the suggestion and squash but happy to go either way if you think there was a reason to keep the old way.